modules/item/src/main/java/unnamed/mmo/item/crafting/CraftingInventory.java

```diff
@@ -0,0 +1,197 @@
+            } else {
+                // Figure out how many we can add
+                for(int i = totalItems; i > 0; i -= outputAmount) {
+                    boolean canAdd = player.getInventory().addItemStack(getItemStack(0).withAmount(i), TransactionOption.DRY_RUN);
+                    if(canAdd) {
+                        player.getInventory().addItemStack(getItemStack(0).withAmount(totalItems), TransactionOption.ALL);
+                        final int craftNum = totalItems / outputAmount;
+                        for (int j = 1; j <= 9; j++) {
+                            // TODO: This does not consider if more than one item is consumed in a craft
+                            if(getItemStack(i).amount() <= craftNum) {
+                                setItemStack(i, ItemStack.AIR);
+                            } else {
+                                setItemStack(i, getItemStack(i).withAmount(integer -> integer - craftNum));
+                            }
+                        }
+                        break;
+                    }
+                }
+            }
+            updateCraftingRecipe();
+            return true;
+        }
+        // And then we copy the code from Inventory#shiftClick, except we start our clickProcessor search at 1 instead to avoid
+        // shift clicking things into the crafting slot
+        final PlayerInventory playerInventory = player.getInventory();
+        final boolean isInWindow = slot <= 9;
+        final int clickSlot = isInWindow ? slot : PlayerInventoryUtils.convertSlot(slot, 10);
+        final ItemStack clicked = isInWindow ? getItemStack(slot) : playerInventory.getItemStack(clickSlot);
+        final ItemStack cursor = getCursorItem(player); // Isn't used in the algorithm
+        final InventoryClickResult clickResult = this.clickProcessor.shiftClick(
+                isInWindow ? this : playerInventory,
+                isInWindow ? playerInventory : this,
+                1, isInWindow ? playerInventory.getInnerSize() : getInnerSize(), 1,
+                player, clickSlot, clicked, cursor);
+        if (clickResult.isCancel()) {
+            updateAll(player);
+            return false;
+        }
+        if (isInWindow) {
+            setItemStack(slot, clickResult.getClicked());
+        } else {
+            playerInventory.setItemStack(clickSlot, clickResult.getClicked());
+        }
+        updateAll(player); // FIXME: currently not properly client-predicted
+        setCursorItem(player, clickResult.getCursor());
+        return true;
+    }
+
+    private void updateCraftingRecipe() {
+        List<ItemStack> currentRecipe = List.of(Arrays.copyOfRange(getItemStacks(), 1, 9));
+        for (CraftingRecipe recipe : recipeList.getRecipeList()) {
+            if (recipe.doesRecipeMatch(currentRecipe)) {
+                setItemStack(0, recipe.getRecipeOutput());
+                return;
+            }
+        }
+        setItemStack(0, ItemStack.AIR);
```

Comment:
this is a mega nitpick, but it would be nice if there were some constants defined in this file for the named slots